### ISSUE TITLE
libigl Viewer crashes in Debug mode in VS2017, due to strange Eigen error of nanogui

### ISSUE THREAD

**Issue by milkpku** [wjakob/nanogui#286]
Hi, dear author,

There is a problem of `igl::viewer::Viewer` in [**libigl**](https://github.com/libigl/libigl) troubling me for a long time, and recently I figured out the error maybe located in **nanogui**. Can you help to fix this?

I compile program with `igl::viewer::Viewer` in VS2015/17, linking it with Release/Debug mode static lib of **nanogui**. The program have no problem under Release mode, but will crash under Debug mode. I looked into the program and found that it is `Viewer.cpp line 256` of **libigl** causes the error.

By looking further, I found it is `colorwheel.cpp line 300` of **nanogui** cause the error, but I can go no further at this line. I have copied the code out and tested, but found it works well. I guess a possible cause of this error may come from the definition of macros in `colorwheel.h line 111`.

The following code is the test code I use to show viewer crash, wish this can help:
```c++
#include <GL/glew.h>
#include <GLFW/glfw3.h>

#pragma comment(lib, "opengl32")
#pragma comment(lib, "glew32")
#pragma comment(lib, "glfw3")

#ifndef _DEBUG
#pragma comment(lib, "nanogui")
#else
#pragma comment(lib, "nanoguid")
#endif

#define IGL_VIEWER_WITH_NANOGUI
#include <igl/viewer/Viewer.h>

int main(int argc, char* argv[])
{
  igl::viewer::Viewer viewer;
  viewer.launch();
}
```

**Comment by svenevs** [comment]
@milkpku I am not the author, but I believe you are encountering a problem with `Eigen`.  I suggest an Eigen problem because line 300 is [this](https://github.com/libigl/nanogui/blob/2c016fc9b7d627cffab0fa97737a70b4171f3d3e/src/colorwheel.cpp#L300).

You may want to try using a different version of Eigen.  Their build system [allows you to bypass the `nanogui/ext/Eigen`](https://github.com/libigl/libigl/blob/94671b5b8b3068e5600a05d8b68d55ea0b3d1664/shared/cmake/CMakeLists.txt#L69-L72) in favor of your own version (provided you download a different version).  Overriding would look something like `cmake .. -DEIGEN_INCLUDE_DIR=/some/path/to/your/installation`.

As an alternative, Libigl is using a [somewhat older version](https://github.com/libigl/libigl/tree/94671b5b8b3068e5600a05d8b68d55ea0b3d1664/external) of NanoGUI.  The things that have changed are mostly just new features, but NanoGUI updated the internal copy of Eigen between then and now.  You could try and go into `libigl/external/nanogui` and execute `git pull` to get the latest version of NanoGUI.

However, there are are [other issues with Windows and Eigen](https://github.com/wjakob/nanogui/issues/258) right now.

It may not be very easy to solve your current predicament :/

There are some other issues with your code.  I doubt IGL requires you to manually include `GL/glew.h>` or `<GLFW/glfw3.h>`.  More specifically, doing so **before** NanoGUI (which means before including IGL headers for you) is a problem.  [The way in which NanoGUI includes OpenGL headers is very specific.](https://github.com/wjakob/nanogui/blob/7f9e0a6ae28ef0ec4ea8c8ab6f96e703e861b245/include/nanogui/opengl.h#L18-L32)  It seems unlikely this would cause your specific problem, but if you include GLFW before NanoGUI gets to, things may not workout correctly.  You also shouldn't be using `glew` at all, NanoGUI uses GLFW and GLAD.

Long story short, if you're using IGL's viewier, you shouldn't need to include any of the GLFW headers directly yourself.

I hope some of those give you some direction.

**Comment by milkpku** [comment]
Hi, @svenevs ,

Firstly, sorry for my delayed reply. I have just get time (get over my laziness) to test what you suggest. 

Secondly, the issue is cause by exactly what you have pointed out, namely the mismatch of different versions of **Eigen** when I compile **nanogui** and my program. After compiling **nanogui** with customized **Eigen**, the problem solved.

Thanks again for your suggestion!
